x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java

```diff
@@ -0,0 +1,575 @@
+        });
+    }
+
+    public void testSnapshotPendingDeletionCannotBeRestored() throws Exception {
+        blockPendingDeletionThenExecute((repository, snapshot, index) -> {
+            ConcurrentSnapshotExecutionException exception = expectThrows(
+                ConcurrentSnapshotExecutionException.class,
+                () -> client().admin().cluster().prepareRestoreSnapshot(repository, snapshot.getName()).setWaitForCompletion(true).get()
+            );
+            assertThat(exception.getMessage(), containsString("cannot restore a snapshot already marked as deleted"));
+        });
+    }
+
+    public void testSearchableSnapshotIsDeletedWhenRepoIsRecreated() throws Exception {
+        mountIndexThenExecute((repository, snapshot, index) -> {
+            try {
+                final Settings.Builder repositorySettings = getRepositorySettings(repository);
+                updateRepositoryReadOnly(repository, true);
+
+                assertAcked(client().admin().indices().prepareDelete(mountedIndex(index)));
+                awaitSnapshotPendingDeletion(snapshot);
+
+                assertAcked(client().admin().cluster().prepareDeleteRepository(repository));
+                awaitSnapshotPendingDeletion(snapshot);
+
+                final String repoName;
+                if (randomBoolean()) {
+                    // re register the repository without verification: the snapshot
+                    // pending deletion logic should try to delete the snapshot based
+                    // on the repository name
+                    repoName = repository;
+                    createRepository(repoName, "mock", repositorySettings, false);
+                } else {
+                    // re register the repository under a different name: the snapshot
+                    // pending deletion logic should try to delete the snapshot based
+                    // on the repository uuid, that is why we force a verification here
+                    repoName = "new_" + repository;
+                    createRepository(repoName, "mock", repositorySettings, true);
+                }
+                awaitNoMoreSnapshotsDeletions();
+
+                expectThrows(
+                    SnapshotMissingException.class,
+                    () -> client().admin().cluster().prepareGetSnapshots(repoName).setSnapshots(snapshot.getName()).get()
+                );
+            } catch (Exception e) {
+                throw new AssertionError(e);
+            }
+        });
+    }
+
+    public void testSearchableSnapshotIsDeletedWithOnGoingRestore() throws Exception {
+        mountIndexThenExecute((repository, snapshot, index) -> {
+            try {
+                final String masterNode = internalCluster().getMasterName();
+                blockMasterOnShardLevelSnapshotFile(repository, getRepositoryData(repository).resolveIndexId(index).getId());
```

Comment:
I think the naming of some of the underlying methods is confusing: `setBlockAndOnWriteShardLevelSnapFiles`, just double checking that the effect we are after is to block on reads here?

Reply:
I opened #82530 to rename the method.

Reply:
(and yes, we only want to block reads here)